File: packages/agent/scanner.go
package main

import (
	"context"
	"os"
	"path/filepath"
	"time"
)

var scanCanceled bool

func CancelScan() {
	scanCanceled = true
}

func ResetCancel() {
	scanCanceled = false
}

func findGitRepos(rootDir string) []string {
	var repos []string
	filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return filepath.SkipDir
		}
		if info.Name() == ".git" && info.IsDir() {
			repos = append(repos, filepath.Dir(path))
			return filepath.SkipDir
		}
		if info.IsDir() && info.Name()[0] == '.' && info.Name() != ".git" {
			return filepath.SkipDir
		}
		if info.IsDir() && info.Name() == "node_modules" {
			return filepath.SkipDir
		}
		return nil
	})
	return repos
}

func scanOneRepo(ctx context.Context, git *GitService, repoPath, machine string) GitRepo {
	status, err := git.GetStatusWithLock(ctx, repoPath)
	name := filepath.Base(repoPath)
	if err != nil {
		errStr := err.Error()
		return GitRepo{
			Name:    name,
			Path:    repoPath,
			Machine: machine,
			Error:   &errStr,
		}
	}
	now := time.Now().UnixMilli()
	commitTimeMs := int64(0)
	if status.LastCommitTime != nil {
		commitTimeMs = *status.LastCommitTime * 1000
	}
	return GitRepo{
		Name:           name,
		Path:           repoPath,
		Branch:         &status.Branch,
		HasChanges:     status.HasChanges,
		Staged:         status.Staged,
		Unstaged:       status.Unstaged,
		Untracked:      status.Untracked,
		Ahead:          status.Ahead,
		Behind:         status.Behind,
		Remote:         status.Remote,
		LastCommitTime: &commitTimeMs,
		WeekCommits:    status.WeekCommits,
		LastScanTime:   &now,
		Machine:        machine,
	}
}

func mergeSettings(repo GitRepo, existingRepos []GitRepo) GitRepo {
	for _, e := range existingRepos {
		if e.Path == repo.Path && e.Settings != nil {
			repo.Settings = e.Settings
			break
		}
	}
	return repo
}

func scanGitReposConcurrently(ctx context.Context, git *GitService, repoPaths []string, machine string, existingRepos []GitRepo) ([]GitRepo, []int) {
	type result struct {
		index int
		repo  GitRepo
	}

	results := make([]GitRepo, len(repoPaths))
	sem := make(chan struct{}, 8)
	resCh := make(chan result, len(repoPaths))

	for i, path := range repoPaths {
		sem <- struct{}{}
		go func(idx int, p string) {
			defer func() { <-sem }()
			repoCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
			defer cancel()
			repo := scanOneRepo(repoCtx, git, p, machine)
			repo = mergeSettings(repo, existingRepos)
			resCh <- result{idx, repo}
		}(i, path)
	}

	for i := 0; i < len(repoPaths); i++ {
		r := <-resCh
		results[r.index] = r.repo
	}

	scannedResults := make([]GitRepo, 0, len(repoPaths))
	fetchable := make([]int, 0)
	for i, repo := range results {
		if repo.Path != "" {
			scannedResults = append(scannedResults, repo)
			if !scanCanceled && repo.Settings != nil && !repo.Settings.SkipPullCheck && !repo.Settings.Hidden {
				fetchable = append(fetchable, i)
			}
		}
	}

	return scannedResults, fetchable
}

func fetchReposConcurrently(ctx context.Context, git *GitService, scannedResults []GitRepo, fetchable []int, progressCh chan<- ScanProgress) {
	if progressCh == nil {
		progressCh = make(chan ScanProgress, 100)
	}

	fetchTotal := len(fetchable)
	fetchSem := make(chan struct{}, 4)
	type fetchResult struct {
		index int
		repo  GitRepo
	}
	fetchRes := make(chan fetchResult, fetchTotal)

	for _, idx := range fetchable {
		if scanCanceled {
			break
		}
		fetchSem <- struct{}{}
		go func(i int) {
			defer func() { <-fetchSem }()

			repo := scannedResults[i]
			select {
			case progressCh <- ScanProgress{Phase: "fetching", Total: fetchTotal, Current: 0, Repo: &repo}:
			default:
			}

			fetchCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
			defer cancel()

			git.RunWithLock(fetchCtx, "fetch origin", repo.Path, 30*time.Second)
			status, _ := git.GetStatusWithLock(ctx, repo.Path)
			if status != nil {
				now := time.Now().UnixMilli()
				commitTimeMs := int64(0)
				if status.LastCommitTime != nil {
					commitTimeMs = *status.LastCommitTime * 1000
				}
				updated := GitRepo{
					Name:           repo.Name,
					Path:           repo.Path,
					Branch:         &status.Branch,
					HasChanges:     status.HasChanges,
					Staged:         status.Staged,
					Unstaged:       status.Unstaged,
					Untracked:      status.Untracked,
					Ahead:          status.Ahead,
					Behind:         status.Behind,
					Remote:         status.Remote,
					LastCommitTime: &commitTimeMs,
					WeekCommits:    status.WeekCommits,
					LastScanTime:   &now,
					Machine:        repo.Machine,
					Settings:       repo.Settings,
				}
				fetchRes <- fetchResult{i, updated}
			} else {
				fetchRes <- fetchResult{i, repo}
			}
		}(idx)
	}

	for i := 0; i < fetchTotal; i++ {
		r := <-fetchRes
		scannedResults[r.index] = r.repo
		select {
		case progressCh <- ScanProgress{Phase: "fetching", Total: fetchTotal, Current: i + 1, Repo: &r.repo}:
		default:
		}
	}
}

func scanAll(ctx context.Context, git *GitService, cache *CacheService, rootDir, machine string, progressCh chan<- ScanProgress) {
	defer close(progressCh)

	repoPaths := findGitRepos(rootDir)
	total := len(repoPaths)

	existingRepos, _ := cache.Load()

	progressCh <- ScanProgress{Phase: "discovering", Total: total}

	scannedResults, fetchable := scanGitReposConcurrently(ctx, git, repoPaths, machine, existingRepos)

	for i, repo := range scannedResults {
		r := repo
		progressCh <- ScanProgress{Phase: "scanning", Total: total, Current: i + 1, Repo: &r}
	}

	if !scanCanceled {
		cache.Save(scannedResults)
	}

	fetchReposConcurrently(ctx, git, scannedResults, fetchable, progressCh)

	if !scanCanceled {
		cache.Save(scannedResults)
	}

	for i, repo := range scannedResults {
		r := repo
		progressCh <- ScanProgress{Phase: "fetching", Total: len(scannedResults), Current: i + 1, Repo: &r}
	}

	progressCh <- ScanProgress{Phase: "done", Total: len(scannedResults), Current: len(scannedResults)}
}

func scanOnly(ctx context.Context, git *GitService, cache *CacheService, rootDir, machine string, progressCh chan<- ScanProgress) {
	defer close(progressCh)

	repoPaths := findGitRepos(rootDir)
	total := len(repoPaths)

	existingRepos, _ := cache.Load()

	progressCh <- ScanProgress{Phase: "discovering", Total: total}

	scannedResults, _ := scanGitReposConcurrently(ctx, git, repoPaths, machine, existingRepos)

	for i, repo := range scannedResults {
		r := repo
		progressCh <- ScanProgress{Phase: "scanning", Total: total, Current: i + 1, Repo: &r}
	}

	if !scanCanceled {
		cache.Save(scannedResults)
	}

	progressCh <- ScanProgress{Phase: "done", Total: len(scannedResults), Current: len(scannedResults)}
}

File: packages/server-bun/scanner.ts
import { readdirSync, statSync } from "fs";
import { join, basename } from "path";
import type { GitRepo, GitRepoSettings, ScanProgress } from "./types";
import type { GitService } from "./git";
import type { CacheService } from "./cache";

let scanCanceled = false;

export function cancelScan(): void {
  scanCanceled = true;
}

export function resetCancel(): void {
  scanCanceled = false;
}

function findGitRepos(rootDir: string): string[] {
  const repos: string[] = [];

  function walk(dir: string): void {
    let entries: string[];
    try {
      entries = readdirSync(dir);
    } catch {
      return;
    }

    for (const entry of entries) {
      const fullPath = join(dir, entry);
      let st;
      try {
        st = statSync(fullPath);
      } catch {
        continue;
      }
      if (!st.isDirectory()) continue;

      if (entry === ".git") {
        repos.push(dir);
        return;
      }
      if (entry.startsWith(".") && entry !== ".git") continue;
      if (entry === "node_modules") continue;

      walk(fullPath);
    }
  }

  walk(rootDir);
  return repos;
}

async function scanOneRepo(
  git: GitService,
  repoPath: string,
  machine: string,
): Promise<GitRepo> {
  const name = basename(repoPath);
  try {
    const status = await git.getStatusWithLock(repoPath);
    const now = Date.now();
    const commitTimeMs = status.lastCommitTime
      ? status.lastCommitTime * 1000
      : 0;
    return {
      name,
      path: repoPath,
      branch: status.branch,
      hasChanges: status.hasChanges,
      staged: status.staged,
      stagedFiles: status.stagedFiles,
      unstaged: status.unstaged,
      unstagedFiles: status.unstagedFiles,
      untracked: status.untracked,
      untrackedFiles: status.untrackedFiles,
      ahead: status.ahead,
      behind: status.behind,
      remote: status.remote,
      lastCommitTime: commitTimeMs,
      weekCommits: status.weekCommits,
      lastScanTime: now,
      machine,
      settings: null,
    };
  } catch (err: any) {
    return {
      name,
      path: repoPath,
      branch: null,
      hasChanges: false,
      staged: 0,
      stagedFiles: [],
      unstaged: 0,
      unstagedFiles: [],
      untracked: 0,
      untrackedFiles: [],
      ahead: 0,
      behind: 0,
      remote: null,
      lastCommitTime: null,
      weekCommits: 0,
      lastScanTime: null,
      error: err.message,
      machine,
      settings: null,
    };
  }
}

function mergeSettings(repo: GitRepo, existingRepos: GitRepo[]): GitRepo {
  for (const e of existingRepos) {
    if (e.path === repo.path && e.settings) {
      return { ...repo, settings: e.settings };
    }
  }
  return repo;
}

async function scanGitReposConcurrently(
  git: GitService,
  repoPaths: string[],
  machine: string,
  existingRepos: GitRepo[],
): Promise<{ results: GitRepo[]; fetchable: number[] }> {
  const results: GitRepo[] = new Array(repoPaths.length);
  const sem = new Array(8).fill(null);
  let semIdx = 0;

  const promises = repoPaths.map(async (path, i) => {
    // Simple semaphore
    while (semIdx >= 8) await new Promise((r) => setTimeout(r, 10));
    semIdx++;
    try {
      const repo = await scanOneRepo(git, path, machine);
      results[i] = mergeSettings(repo, existingRepos);
    } finally {
      semIdx--;
    }
  });

  await Promise.all(promises);

  const scannedResults: GitRepo[] = [];
  const fetchable: number[] = [];
  for (let i = 0; i < results.length; i++) {
    const repo = results[i];
    if (repo.path) {
      scannedResults.push(repo);
      if (
        !scanCanceled &&
        repo.settings &&
        !repo.settings.skipPullCheck &&
        !repo.settings.hidden
      ) {
        fetchable.push(i);
      }
    }
  }

  return { results: scannedResults, fetchable };
}

async function fetchReposConcurrently(
  git: GitService,
  scannedResults: GitRepo[],
  fetchable: number[],
  progressCh: (p: ScanProgress) => void,
): Promise<void> {
  const fetchTotal = fetchable.length;
  const fetchSem = new Array(4).fill(null);
  let fetchSemIdx = 0;

  const promises = fetchable.map(async (idx) => {
    if (scanCanceled) return;

    while (fetchSemIdx >= 4) await new Promise((r) => setTimeout(r, 10));
    fetchSemIdx++;
    try {
      const repo = scannedResults[idx];
      progressCh({
        phase: "fetching",
        total: fetchTotal,
        current: 0,
        repo,
      });

      await git.runWithLock(["fetch", "origin"], repo.path, 30_000);
      const status = await git.getStatusWithLock(repo.path).catch(() => null);

      if (status) {
        const now = Date.now();
        const commitTimeMs = status.lastCommitTime
          ? status.lastCommitTime * 1000
          : 0;
        scannedResults[idx] = {
          ...repo,
          branch: status.branch,
          hasChanges: status.hasChanges,
          staged: status.staged,
          stagedFiles: status.stagedFiles,
          unstaged: status.unstaged,
          unstagedFiles: status.unstagedFiles,
          untracked: status.untracked,
          untrackedFiles: status.untrackedFiles,
          ahead: status.ahead,
          behind: status.behind,
          remote: status.remote,
          lastCommitTime: commitTimeMs,
          weekCommits: status.weekCommits,
          lastScanTime: now,
        };
      }

      progressCh({
        phase: "fetching",
        total: fetchTotal,
        current: idx + 1,
        repo: scannedResults[idx],
      });
    } finally {
      fetchSemIdx--;
    }
  });

  await Promise.all(promises);
}

export async function scanAll(
  git: GitService,
  cache: CacheService,
  rootDir: string,
  machine: string,
  progressCh: (p: ScanProgress) => void,
): Promise<void> {
  const repoPaths = findGitRepos(rootDir);
  const total = repoPaths.length;
  const existingRepos = cache.load();

  progressCh({ phase: "discovering", total, current: 0, repo: null });

  const { results: scannedResults, fetchable } =
    await scanGitReposConcurrently(git, repoPaths, machine, existingRepos);

  for (let i = 0; i < scannedResults.length; i++) {
    progressCh({
      phase: "scanning",
      total,
      current: i + 1,
      repo: scannedResults[i],
    });
  }

  if (!scanCanceled) {
    cache.save(scannedResults);
  }

  await fetchReposConcurrently(git, scannedResults, fetchable, progressCh);

  if (!scanCanceled) {
    cache.save(scannedResults);
  }

  for (let i = 0; i < scannedResults.length; i++) {
    progressCh({
      phase: "fetching",
      total: scannedResults.length,
      current: i + 1,
      repo: scannedResults[i],
    });
  }

  progressCh({
    phase: "done",
    total: scannedResults.length,
    current: scannedResults.length,
    repo: null,
  });
}

export async function scanOnly(
  git: GitService,
  cache: CacheService,
  rootDir: string,
  machine: string,
  progressCh: (p: ScanProgress) => void,
): Promise<void> {
  const repoPaths = findGitRepos(rootDir);
  const total = repoPaths.length;
  const existingRepos = cache.load();

  progressCh({ phase: "discovering", total, current: 0, repo: null });

  const { results: scannedResults } = await scanGitReposConcurrently(
    git,
    repoPaths,
    machine,
    existingRepos,
  );

  for (let i = 0; i < scannedResults.length; i++) {
    progressCh({
      phase: "scanning",
      total,
      current: i + 1,
      repo: scannedResults[i],
    });
  }

  if (!scanCanceled) {
    cache.save(scannedResults);
  }

  progressCh({
    phase: "done",
    total: scannedResults.length,
    current: scannedResults.length,
    repo: null,
  });
}
